hathor/feature_activation/feature_service.py

```diff
@@ -92,32 +91,33 @@ def _calculate_new_state(
             return FeatureState.DEFINED
 
         if previous_state is FeatureState.STARTED:
-            if height >= criteria.timeout_height and not criteria.activate_on_timeout:
+            if height >= criteria.timeout_height and not criteria.lock_in_on_timeout:
                 return FeatureState.FAILED
 
-            if (
-                height >= criteria.timeout_height
-                and criteria.activate_on_timeout
-                and height >= criteria.minimum_activation_height
-            ):
-                return FeatureState.ACTIVE
-
             # Get the count for this block's parent. Since this is a boundary block, its parent count represents the
             # previous evaluation interval count.
             parent_block = boundary_block.get_block_parent()
             counts = parent_block.get_feature_activation_bit_counts()
             count = counts[criteria.bit]
             threshold = criteria.get_threshold(self._feature_settings)
 
-            if (
-                height < criteria.timeout_height
-                and count >= threshold
-                and height >= criteria.minimum_activation_height
-            ):
-                return FeatureState.ACTIVE
+            if height < criteria.timeout_height and count >= threshold:
+                return FeatureState.LOCKED_IN
+
+            if (height + evaluation_interval >= criteria.timeout_height) and criteria.lock_in_on_timeout:
+                return FeatureState.MUST_SIGNAL
 
             return FeatureState.STARTED
 
+        if previous_state is FeatureState.MUST_SIGNAL:
+            return FeatureState.LOCKED_IN
```

Comment:
Shouldn't we check the block height to change to LOCKED_IN?

Reply:
No, because the `MUST_SIGNAL` state is defined to always take exactly one evaluation interval. Since this method is only called for boundary blocks, it is guaranteed that after exactly one evaluation interval in `MUST_SIGNAL`, the feature will transition to `LOCKED_IN`, as it should. Any blocks _inside_ that evaluation interval, that is, between the boundary blocks, will continue to be in `MUST_SIGNAL` since they will not call this method and just propagate their parent's state instead.

But just to be clear, what was your thought reasoning?

Reply:
I was concerned that this method might be called multiple times. I feel it's worth adding a comment about what you said here to clarify for future code readers.

Reply:
I just checked the assert for boundary blocks. I feel it's worth adding a note about it on the docstring.

Reply:
Sure! I updated the docstring and added a comment in https://github.com/HathorNetwork/hathor-core/pull/675/commits/a91c88ca00cc933c38d653849c346c73466e8906

Reply:
Great! Thanks!